### PR TITLE
Support unzipped entry

### LINKED ISSUE #215: [Feature Request] Support Chapters that are not in a comic book file
**Is your feature request related to a problem? Please describe.**

I am running Mango on a VPS that has a limited amount of storage, but I have 1TB OneDrive, so I mounted it with `rclone` and changed the `docker-compose.yml`:
```diff,yaml
version: '3'

services:
  mango:
    image: hkalexling/mango
    container_name: mango
    ports:
      - 9005:9000
    volumes:
      - ./mango:/root/mango
      - ./config/mango:/root/.config/mango
+     - /mnt/onedrive/manga:/root/mango/library
```

After I did this, the website became really slow and scanning a single manga with 2 Chapters (few MBs) took 27300ms to scan, which seems to be caused by #118.

It seems like `rclone` is downloading the entire library for mango, which is quite expensive and kinda defeats the purpose of storing them in the cloud.

**Describe the solution you'd like**
Currently, Mango only supports libraries that look like this:
```
library
├── Manga Name
│  ├── Chapter1.cbz
│  ├── Chapter2.cbz
│  └── ....
└── ....
```

Instead of having the chapters in zip files, one should be able to store them in directories:

```
library
├── Manga Name
│  ├── Chapter1
│  │  ├── 001.png
│  │  └── ....
│  └── ....
└── ....
```

This should prevent `rclone` from downloading every single zip in the library, just to generate the preview images.

It might be good to abstract away the zip under an interface, so directories and comic book files are treated in the same way.
This would have the benefit that comic book files that contain the chapters as directories would be supported as well.

### PR DESCRIPTION
**Pull request by Leeingnyo**
Resolve #215

## What I did

* Change the `Entry` class to abstract class,
* Implement `DirectoryEntry`, which treats unzipped images in a directory as an entry
  * the rule is described at [here](https://github.com/hkalexling/Mango/issues/215#issuecomment-926544760)
* Scan `DirectoryEntry` when `Title#new` and `Title#examine` are called.
* Fix caching policy after some tests

## Result and test

Here is an test library environment.

```
.:
 info.json   nested-folder/  'THE iDOLM@STER Your Mess@ge 3rd stage.zip'  'THE iDOLM@STER Your Mess@ge 4th stage.zip'   unzipped-files/

./nested-folder:
 10.jpg   12.jpg   14.jpg   16.jpg   18.jpg   1.jpg    21.jpg   23.jpg   25.jpg   2.jpg   4.jpg   6.jpg   8.jpg   info.json             'THE iDOLM@STER Your Mess@ge 1st stage.zip'
 11.jpg   13.jpg   15.jpg   17.jpg   19.jpg   20.jpg   22.jpg   24.jpg   26.jpg   3.jpg   5.jpg   7.jpg   9.jpg   nested-unziped-file/  'THE iDOLM@STER Your Mess@ge 2nd stage.zip'

./nested-folder/nested-unziped-file:
10.jpg  11.jpg  12.jpg  13.jpg  14.jpg  15.jpg  16.jpg  17.jpg  18.jpg  19.jpg  1.jpg  20.jpg  21.jpg  22.jpg  23.jpg  24.jpg  25.jpg  26.jpg  2.jpg  3.jpg  4.jpg  5.jpg  6.jpg  7.jpg  8.jpg  9.jpg

./unzipped-files:
10.jpg  11.jpg  12.jpg  13.jpg  14.jpg  15.jpg  16.jpg  17.jpg  18.jpg  19.jpg  1.jpg  20.jpg  21.jpg  22.jpg  23.jpg  24.jpg  25.jpg  26.jpg  2.jpg  3.jpg  4.jpg  5.jpg  6.jpg  7.jpg  8.jpg  9.jpg
```

In tree view,

```
root (unzipped, have 4 entries, one title)
├ nested-folder (would be an entry and a title, have 3 entries)
│ ├ 1.jpg ~ 26.jpg
│ ├ nested-unziped-file (would be an entry)
│ │ └ 1.jpg ~ 26.jpg
│ ├ THE iDOLM@STER Your Mess@ge 1st stage.zip
│ └ THE iDOLM@STER Your Mess@ge 2nd stage.zip
├ unzipped-files (would be an entry)
│ └ 1.jpg ~ 26.jpg
├ THE iDOLM@STER Your Mess@ge 3rd stage.zip
└ THE iDOLM@STER Your Mess@ge 4th stage.zip
```

### Result

image

image

image

The directories appeared as entries and titles

### What I tested

* remove an image file from directory
* rename an image file in a directory
* add new folder to scan

## About class name

ZippedEntry -> ArchiveEntry
DirectoryEntry -> ?

I can't come up with good names...

### CODE REVIEW

**Comment by Leeingnyo** [comment]
I found that there's a bug after rename a directory entry and scan.

**Comment by Leeingnyo** [comment]
@tr7zw Hi!

you mean that the `nested-folder` in the sample would be a problem, and it shouldn't be treated as an entry because it has another titles and zipped entries though it has image files, right?
I think the responsibility of configuring a library is up to the library owners. If they don't want this, they would remove those files.

**Comment by tr7zw** [comment]
Huh I guess never mind? I still had in mind that if you have a folder "Manga" and in that folder a folder "One Piece" with the chapters inside, I think in the past it treated the "Manga" folder as something containing Chapters. But apparently this was already fixed, now it shows Titles/Entries. 
Still a really neat pr, since using directories as chapters allows some more flexibility, and with the new abstract Entry class it could allow some nice new features(other compressed files than .zip, dynamically loading chapters from urls/network shares).

**Comment by hkalexling** [comment]
Thanks @Leeingnyo! I fixed the linter warnings and took a quick look, and it looks good overall! I think it's a nice application of abstract classes.

Re class names, yeah I think `ArchiveEntry` is more appropriate than `ZippedEntry` because we also support RAR/CBR files. I think `DirectoryEntry` looks fine. We could also use `DirEntry` to make it shorter but I have no strong feelings.

Also I think it makes more sense to break the classes into individual files, e.g., `entry.cr`, `zipped_entry.cr`, and `directory_entry.cr`. What do you think?

I will do a full review and some testings later this week :+1:

**Review by hkalexling — APPROVED** [review]

**Comment by hkalexling** [comment]
Hey @Leeingnyo sorry for the delay. I made quite some changes to the PR. Can you take a quick look when you have the time to make sure I didn't accidentally fuck up anything? Thanks!

File: src/routes/api.cr [hunk 1]
```diff
@@ -40,7 +40,7 @@ struct APIRouter
     Koa.schema "entry", {
       "pages" => Int32,
       "mtime" => Int64,
-    }.merge(s %w(zip_path title size id title_id display_name cover_url)),
+    }.merge(s %w(path title size id title_id display_name cover_url)),
```
[the remark this answers is not in the document]
Comment (Leeingnyo) [reply src/routes/api.cr]: Fixed!

**Comment by Leeingnyo** [comment]
@hkalexling sorry I pushed errors occurred version... wait for a moment

**Comment by Leeingnyo** [comment]
@hkalexling it looks great. I like the method that you did to recover entry instances :)
Since `@page` was added to `Entry`, it caused an error to recover old `library.yml.zip` file. But, I don't mind of this. fine!

**Comment by hkalexling** [comment]
Thanks @Leeingnyo! Sorry the comments above was just for my own reference and I accidentally published them ;-P

> Since @page was added to Entry, it caused an error to recover old library.yml.zip file.

Could you elaborate a bit on this? I tried the following steps but didn't see the error.

1. Remove `library.yml.gz`
2. Build and run from the `master` branch to regenerate the `library.yml.gz` file
3. Build and run from this branch and there's no error

**Comment by Leeingnyo** [comment]
Oh that's the difference. I built it from the `dev` branch. but I had the same error when I tried from the `master` branch.

```
[ERROR]   2022/06/05 15:02:04 | Missing YAML attribute: path at line 14, column 7
```

because my ubuntu snap upgrades a Crystal implicitly, I use a Crystal 1.4.1... this would be a matter. :p

**Comment by hkalexling** [comment]
Ah sorry my bad. It does happen on `master` on Crystal `1.0.0` as well, but it's a minor issue and will only happen once so I think we are fine.

**Comment by hkalexling** [comment]
Thanks <3

**Pull request merged**